compiler/hash-attrs/src/target.rs

```diff
@@ -0,0 +1,292 @@
+        const ModDef = 1 << 10;
+
+        /// A block, specifically the [`hash_ast::ast::Block::Body`] variant.
+        const Block = 1 << 11;
+
+        /// An `import` statement.
+        const Import = 1 << 12;
+
+        /// A type function definition.
+        const TyFnDef = 1 << 13;
+
+        /// A `struct` definition.
+        const StructDef = 1 << 14;
+
+        /// An `enum` definition.
+        const EnumDef = 1 << 15;
+
+        /// An enum variant.
+        const EnumVariant = 1 << 16;
+
+        /// A function definition, regardless of the position.
+        const FnDef = 1 << 17;
+
+        /// A type.
+        const Ty = 1 << 18;
+
+        /// A type argument, whilst this has similar implications to
+        /// a type, it adds additional context that this is a type
+        /// argument.
+        const TyArg = 1 << 19;
+
+        /// A parameter in a struct, enum, tuple or function definition.
+        const Param = 1 << 20;
+
+        /// A field in a struct or enum.
+        const Field = 1 << 21;
+
+        /// A match branch.
+        const MatchCase = 1 << 22;
+
+        /// A pattern.
+        const Pat  = 1 << 23;
+
+        /// A pattern argument.
+        const PatArg  = 1 << 24;
+
+        /// A general item definition e.g. `struct`, `enum`, `impl`, `mod` and `fn`.
+        const Item = Self::StructDef.bits() | Self::EnumDef.bits() | Self::FnDef.bits() | Self::TyFnDef.bits() | Self::ImplDef.bits() | Self::ModDef.bits();
+    }
+}
+
+impl AttrTarget {
+    /// Classify the given [ast::Expr] into a [AttrTarget].
+    pub fn classify_expr(expr: &ast::Expr) -> Self {
+        match expr {
+            ast::Expr::ConstructorCall(_) => AttrTarget::ConstructorCall,
+            ast::Expr::Macro(_) => AttrTarget::MacroInvocation,
+            ast::Expr::Unsafe(_) => AttrTarget::Unsafe,
+            ast::Expr::Lit(_) => AttrTarget::Lit,
+            ast::Expr::Block(ast::BlockExpr { data: block }) => match block.body() {
+                ast::Block::Loop(_) | ast::Block::While(_) | ast::Block::For(_) => AttrTarget::Loop,
+                ast::Block::Match(_) | ast::Block::If(_) => AttrTarget::Match,
+                ast::Block::Body(_) => AttrTarget::Block,
+            },
+            ast::Expr::Import(_) => AttrTarget::Import,
+            ast::Expr::StructDef(_) => AttrTarget::StructDef,
+            ast::Expr::EnumDef(_) => AttrTarget::EnumDef,
+            ast::Expr::TyFnDef(_) => AttrTarget::TyFnDef,
+            ast::Expr::TraitDef(_) => AttrTarget::Item,
+            ast::Expr::ImplDef(_) => AttrTarget::ImplDef,
+            ast::Expr::ModDef(_) => AttrTarget::ModDef,
+            ast::Expr::FnDef(_) => AttrTarget::FnDef,
+            ast::Expr::Ty(_) => AttrTarget::Ty,
+
+            // If this is a declaration, we have to recurse into the subject...
+            ast::Expr::Declaration(ast::Declaration { value: Some(value), .. }) => {
+                AttrTarget::classify_expr(value.body())
```

Comment:
What if someone wants to put an attribute on the declaration itself?

Reply:
Im unsure how to solve this problem.

Reply:
Maybe it should always apply on the declaration, and macros should handle this accordingly?

Reply:
This causes problems because the `AstNodeId` that the attributes are then registered on isn't used to determine whether it has attributes on it. E.g. when you write:

```rs
#[link_name("foo")]
foo := () => { ... }
```

Lowering and codegen use the `AstNodeId` of the `() => { ... }` rather than the declaration itself whether it specifies any attributes.

Reply:
That's possible, but it's a bit of a hack.. I guess we could argue that declarations are not "real" expressions so macros never apply to them, they always apply to the inner expressions.. Anyway I think it should be fine to keep it like this for now and if we encounter a use case of putting a macro on a declaration we can reconsider. Thoughts?

Reply:
Yeah, I'm gonna make this thread into an issue/discussion so that we can come back to this later and properly decide.